File: Flatten Binary Tree to Linked List/Flatten Binary Tree to Linked List.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> stk = new Stack();
        if(root != null){
            stk.push(root);
        }
        while(!stk.empty()){
            TreeNode node = stk.pop();
            if(node.right != null){
                stk.push(node.right);
            }
            if(node.left != null){
                stk.push(node.left);
                node.right = node.left;
                node.left = null;
            }
            else{
                if(stk.empty()){
                    node.right = null;
                }
                else{
                    node.right = stk.peek();
                }
            }
        }
    }
}

File: Flatten Binary Tree to Linked List/Flatten Binary Tree to Linked List.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    void flatten(TreeNode *root) {
        if(!root)       return;
        
        std::stack<TreeNode*> stk;
        stk.push(root);
        while(!stk.empty()){
            TreeNode* curr = stk.top();
            stk.pop();
            if(curr->left){
                if(curr->right){
                    stk.push(curr->right);
                }
                curr->right = curr->left;
                stk.push(curr->left);
                curr->left = NULL;
            }
            else if(curr->right){
                stk.push(curr->right);
            }
            else{
                if(stk.empty()){
                    curr->right = NULL;
                }
                else{
                    curr->right = stk.top();
                }
            }
        }
        return;
    }
};
